docs/hooks.rst

```diff
@@ -1,43 +1,312 @@
+main loop.  A partial list of some of the more important ones is presented here:
+
+- **intro**: *str* - if provided this serves as the intro banner printed once
+  at start of application, after ``preloop`` runs
+- **allow_cli_args**: *bool* - if True (default), then searches for -t or
+  --test at command line to invoke transcript testing mode instead of a normal
+  main loop and also processes any commands provided as arguments on the
+  command line just prior to entering the main loop
+- **echo**: *bool* - if True, then the command line entered is echoed to the
+  screen (most useful when running scripts)
+- **prompt**: *str* - sets the prompt which is displayed, can be dynamically
+  changed based on application state and/or command results
+
+
+Command Processing Loop
+-----------------------
+
+When you call `.cmdloop()`, the following sequence of events are repeated until
+the application exits:
+
+1. Output the prompt
+2. Accept user input
+3. Parse user input into `Statement` object
+4. Call methods registered with `register_postparsing_hook()`
+5. Call `postparsing_precmd()` - for backwards compatibility with prior releases of cmd2, now deprecated
+6. Redirect output, if user asked for it and it's allowed
+7. Start timer
+8. Call methods registered with `register_precmd_hook()`
+9. Call `precmd()` - for backwards compatibility with ``cmd.Cmd``
+10. Add statement to history
+11. Call `do_command` method
+12. Call methods registered with `register_postcmd_hook()`
+13. Call `postcmd(stop, statement)` - for backwards compatibility with ``cmd.Cmd``
+14. Stop timer and display the elapsed time
+15. Stop redirecting output if it was redirected
+16. Call methods registered with `register_cmdfinalization_hook()`
+17. Call `postparsing_postcmd()` - for backwards compatibility - deprecated
+
+By registering hook methods, steps 4, 8, 12, and 16 allow you to run code
+during, and control the flow of the command processing loop. Be aware that
+plugins also utilize these hooks, so there may be code running that is not part
+of your application. Methods registered for a hook are called in the order they
+were registered. You can register a function more than once, and it will be
+called each time it was registered.
+
+Postparsing, precommand, and postcommand hook methods share some common ways to
+influence the command processing loop.
+
+If a hook raises a ``cmd2.EmptyStatement`` exception:
+- no more hooks (except command finalization hooks) of any kind will be called
+- if the command has not yet been executed, it will not be executed
+- no error message will be displayed to the user
+
+If a hook raises any other exception:
+- no more hooks (except command finalization hooks) of any kind will be called
+- if the command has not yet been executed, it will not be executed
+- the exception message will be displayed for the user.
+
+Specific types of hook methods have additional options as described below.
+
+Postparsing Hooks
```

Comment:
If you decide to add a list of preparsing hooks, then add a section for that before this.

Reply:
See above rationale for not having preparsing hooks.